## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [ ]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [ ]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [ ]:
# Use the describe function to gather some basic statistics
print(df.describe())

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [ ]:
# Use this space to do any additional research
# and familiarize yourself with the data.
# Load the dataset




In [ ]:
# What three item categories had the most entries?
top_categories=df['category'].value_counts().head(3)
print("Top Categories:\n",top_categories)

Top Categories:
 category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [ ]:
# For the category with the most entries, which subcategory had the most entries?
top_categories=top_categories.index[0]
top_subcategories =df[df['category']==top_categories]['subcategory'].value_counts().head(1)
print("Top Subcatergory:\n",top_subcategories)

Top Subcatergory:
 subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [ ]:
# Which five clients had the most entries in the data?
top_5_ids= df['client_id'].value_counts().head(5)
print("Top 5 cleints:\n",top_5_ids)

Top 5 cleints:
 client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [ ]:
# Store the client ids of those top 5 clients in a list.
top_5_list=top_5_ids.index.tolist()
print("Top client ID:", top_5_list)




Top client ID: [33615, 66037, 46820, 38378, 24741]


In [ ]:
# How many total units (the qty column) did the client with the most entries order order?
most_order=top_5_list[0]
total_unit=df[df['client_id'] == most_order]['qty'].sum()
print("Total unit of client:",total_unit)

Total unit of client: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [ ]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtoatal'] = df['unit_price']*df['qty']

In [ ]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
# Create a column that calculates the subtotal for each line
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7*x if x > 50 else 10*x)



In [ ]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price']=df['subtoatal']+df['shipping_price']
df['total_cost']=df['total_price']*1.0925

In [ ]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost']=df['unit_cost']*df['qty']+df['shipping_price']


In [ ]:
# Create a column for the profit of each line using line cost and line price
df['profit']=df['total_cost']-df['line_cost']


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [ ]:
# Check your work using the totals above
order_2742071 = df[df['order_id'] == 2742071]['total_cost'].sum()
order_2173913 = df[df['order_id'] == 2173913]['total_cost'].sum()
order_6128929 = df[df['order_id'] == 6128929]['total_cost'].sum()

print(f"inv# 2742071 Total : {order_2742071}")
print(f"inv# 2173913 Total: {order_2173913}")
print(f"inv# 6128929 Total: {order_6128929}")

inv# 2742071 Total : 139844.51202499997
inv# 2173913 Total: 119838.35704999999
inv# 6128929 Total: 752748.7454750001


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [ ]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
client_spend = df[df['client_id'].isin(top_client_ids)].groupby('client_id')['total_cost'].sum()
print("Total Revenue for Top 5 Clients:", client_spend)


Total Revenue for Top 5 Clients: client_id
24741    7.667848e+07
33615    6.392918e+06
38378    9.172866e+06
46820    8.006333e+06
66037    8.746255e+06
Name: total_cost, dtype: float64


In [ ]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = df[df['client_id'].isin(top_5_list)].groupby('client_id').agg({'qty': 'sum','shipping_price':'sum','total_cost':'sum','profit':'sum'
}).reset_index()



In [ ]:
# Format the data and rename the columns to names suitable for presentation.
summary_df.columns = ['cleint_id','Total purchased','Total shipping cost','Total rev','total profit']
# Define the money columns. 


# Define a function that converts a dollar amount to millions.
def to_millions(x): return x /1_000_000

# Apply the currency_format_millions function to only the money columns. 


summary_df['shipping_price'] = summary_df['shipping_price'].apply(to_millions)
summary_df['total rev']= summary_df['total shippinf cost'].apply(to_millions)
summary_df['profit'] = summary_df['profit'].apply(to_millions)

# Rename the columns to reflect the change in the money format. 
summary_df.columns = ['Client ID', 'Total Units Purchased', 'Total Shipping Cost (in millions)', 'Total Revenue (in millions)', 'Total Profit (in millions)']




KeyError: 'shipping_price'

In [ ]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
summary_df = summary_df.sort_values(by='profit', ascending=False)

print("Summary DataFrame:", summary_df)

KeyError: 'profit'

In [ ]:
summary= " The reserch indicated the top 5 cleints that had the highest revenue that summarizes the cleints analytics of profibility "
print(summary)

 The reserch indicated the top 5 cleints that had the highest revenue that summarizes the cleints analytics of profibility 
